### PR TITLE
Reduce queries to storage to serve 404 pages

### PR DESCRIPTION
**Pull request by stsewd**
This makes use of `version.documentation_type` to stop guessing if for `404/index.html` vs `404.html` and guessing when to test for `dir/index.html` or `dir/README.html`

- Worst case for html projects: 2 calls to storage (before it was 6)
- Worst case for dirhtml projects: 4 calls to storage (before it was 6)
- Best case for html projects: 1 call to storage (before it was 3)
- Best case for dirhtml projects: 3  calls to storage (before it was 6)

Now we have 2 queries to the db, but I guess that's better than having 2 calls to storage :)

### CODE REVIEW

**Comment by stsewd** [comment]
BTW, I saw that mkdocs projects can be `html` projects too https://www.mkdocs.org/user-guide/configuration/#use_directory_urls. I'm going to add that type in another PR, no changes on this PR or the config file are required for that.

File: readthedocs/proxito/views/serve.py [hunk 1]
```diff
@@ -259,11 +264,17 @@ def get(self, request, proxito_path, template_name='404.html'):
         # If that doesn't work, attempt to serve the 404 of the current version (version_slug)
         # Secondly, try to serve the 404 page for the default version
         # (project.get_default_version())
-        versions = [version_slug]
+        versions = [(version_slug, doc_type)]
         default_version_slug = final_project.get_default_version()
         if default_version_slug != version_slug:
-            versions.append(default_version_slug)
-        for version_slug_404 in versions:
+            doc_type = (
```
Comment (humitos) [inline readthedocs/proxito/views/serve.py:270]: can we change the name of this variable? we are overriding the original one.

```suggestion
            default_version_doc_type = (
```

File: readthedocs/proxito/views/serve.py [hunk 2]
```diff
@@ -259,11 +264,17 @@ def get(self, request, proxito_path, template_name='404.html'):
...
+                Version.objects.filter(project=final_project, slug=default_version_slug)
+                .values_list('documentation_type', flat=True)
+                .first()
+            )
+            versions.append((default_version_slug, doc_type))
```
Comment (humitos) [inline readthedocs/proxito/views/serve.py:275]:
```suggestion
            versions.append((default_version_slug, default_version_doc_type))
```

File: readthedocs/proxito/views/serve.py [hunk 3]
```diff
@@ -259,11 +264,17 @@ def get(self, request, proxito_path, template_name='404.html'):
...
+
+        for version_slug_404, doc_type in versions:
```
Comment (humitos) [inline readthedocs/proxito/views/serve.py:277]: Same here

```suggestion
        for version_slug_404, doc_type_404 in versions:
```

File: readthedocs/proxito/views/serve.py [hunk 4]
```diff
@@ -273,17 +284,18 @@ def get(self, request, proxito_path, template_name='404.html'):
...
-                if storage.exists(storage_filename_path):
-                    log.info(
-                        'Serving custom 404.html page: [project: %s] [version: %s]',
-                        final_project.slug,
-                        version_slug_404,
-                    )
-                    resp = HttpResponse(storage.open(storage_filename_path).read())
-                    resp.status_code = 404
-                    return resp
+            is_htmldir_type = doc_type in (SPHINX_HTMLDIR, MKDOCS)
```
Comment (humitos) [inline readthedocs/proxito/views/serve.py:287]:
```suggestion
            is_htmldir_type = doc_type_404 in (SPHINX_HTMLDIR, MKDOCS)
```

File: readthedocs/proxito/views/serve.py [hunk 5]
```diff
@@ -195,35 +195,40 @@ def get(self, request, proxito_path, template_name='404.html'):
...
-                log.info(
-                    'Redirecting to index file: project=%s version=%s, storage_path=%s',
+        doc_type = (
+            Version.objects.filter(project=final_project, slug=version_slug)
+            .values_list('documentation_type', flat=True)
+            .first()
+        )
+        is_htmldir_type = doc_type in (SPHINX_HTMLDIR, MKDOCS)
+        if is_htmldir_type:
+            for tryfile in ('index.html', 'README.html'):
```
Comment (humitos) [inline readthedocs/proxito/views/serve.py:205]: Why we don't do this in the `ServeBase.get` instead?

I mean, if we already know that this project `is_htmldir_type=True`, there is no reason to fail the first request. We can just serve the `*/index.html` in that case.

Then, we can handle _only_ the `README.html` as an special case in the 404 handler.

File: readthedocs/proxito/views/serve.py (hunk 1 above)
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:270]: We don't use the original variable after those lines

File: readthedocs/proxito/views/serve.py (hunk 5 above)
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:205]: We can't do that,  we serve files other than html. Like `foo/bar.img`, we'll try to serve `foo/bar.img/index.html`.

File: readthedocs/proxito/views/serve.py (hunk 1 above)
Comment (humitos) [reply readthedocs/proxito/views/serve.py:270]: Yes, but it's confusing. It's safer to use a different name that reflects what it is exactly.

File: readthedocs/proxito/views/serve.py (hunk 5 above)
Comment (humitos) [reply readthedocs/proxito/views/serve.py:205]: Actually, aren't we already doing this in the `ServeBase.get` at https://github.com/readthedocs/readthedocs.org/blob/0e3df509e7810e46603be47d268273c596e68455/readthedocs/proxito/views/serve.py#L127-L132 ?

We are adding `index.html` if the URL ends with `/`, so it seems there is no need to try the same again in the 404 handler, right?

Comment (stsewd) [reply readthedocs/proxito/views/serve.py:205]: We are only adding the index file if path ends with `/`, the 404 tries to serve the index from paths not ending with `/`

File: readthedocs/proxito/views/serve.py [hunk 6]
```diff
@@ -195,35 +195,40 @@ def get(self, request, proxito_path, template_name='404.html'):
...
-            if storage.exists(storage_filename_path):
-                log.info(
-                    'Redirecting to index file: project=%s version=%s, storage_path=%s',
+        doc_type = (
+            Version.objects.filter(project=final_project, slug=version_slug)
+            .values_list('documentation_type', flat=True)
+            .first()
+        )
+        is_htmldir_type = doc_type in (SPHINX_HTMLDIR, MKDOCS)
+        if is_htmldir_type:
```
Comment (ericholscher) [inline readthedocs/proxito/views/serve.py:204]: I don't think this check is correct. We have users who want to serve README.html as their index, regardless of their doctype, I don't think we can skip that check. 

I'd like to remove the README check all together though, but I don't think this is the way to do it.

Comment (ericholscher) [reply readthedocs/proxito/views/serve.py:204]: Actually, I don't think we can skip the `index.html` check here either. It doesn't depend on htmldir either:

For example we want `https://docs.readthedocs.io/en/latest` to redirect to `https://docs.readthedocs.io/en/latest/`, and it doesn't matter what doctype it is.

Comment (stsewd) [reply readthedocs/proxito/views/serve.py:204]: I thought we did that here

https://github.com/readthedocs/readthedocs.org/blob/33049ccfc70c335d477348788319879721bec520/readthedocs/proxito/views/serve.py#L125-L126

But looks like we send them to the 404 handler

https://github.com/readthedocs/readthedocs.org/blob/33049ccfc70c335d477348788319879721bec520/readthedocs/proxito/urls.py#L122-L133

Is there a reason why we send them to the 404 view and not just redirect them with `/`? (same for subprojects, translations, single page, etc)

But still I can see cases where users want to serve an index without caring about the doctype. That could be, users using a `html`  doctype and having a directory with an index...

I guess we could at least assume `404.html` for `html` types and `404/index.html` for `htmldir` types when serving a custom 404 page? that will give us one check less

Comment (ericholscher) [reply readthedocs/proxito/views/serve.py:204]: We have to do a query to cloud storage to know if the file exists before we redirect to `/`, and we don't want to do any queries in the normal proxito handler to make it really fast. The point of the 404 handler is to be slower and handle the queries. 

> I guess we could at least assume 404.html for html types and 404/index.html for htmldir types when serving a custom 404 page? that will give us one check less

Yea, that seems easy enough. Though I could see users hacking in a top-level 404.html` even with a `dirhtml` builder. We should probably keep stats, but I assume we almost never serve `404/index.html`, and could consider removing it.

Comment (stsewd) [reply readthedocs/proxito/views/serve.py:204]: https://www.mkdocs.org/user-guide/deploying-your-docs/#404-pages

Looks like mkdocs always outputs a `404.html` file, not a `404/inde.html` file. And github pages expects a `404.html` https://help.github.com/en/github/working-with-github-pages/creating-a-custom-404-page-for-your-github-pages-site

Sphinx with `htmldir` is the only one that outputs a `404/index.html` file.

**Comment by stsewd** [comment]
Ok, I just reverted this PR to only check for `404.html` using the doctype information, this is only check for `404/inde.html` on sphinx htmldir projects, the rest of the builders (even mkdocs) always output a `404.html` file.

This reduces 2 calls for all builders tha aren't `sphinx htmldir`.

**Pull request merged**